### ISSUE TITLE
Road to 1.0 (Milestone I)

### ISSUE THREAD

**Issue by Garnek0** [Garnek0/GarnOS#2]
This project is currently on hold since i dont have a lot of time to work on it due to school. I will try to sneak in some work from time to time though, but its probably going to be (very) slow progress until the end of the school year.

### CPU
- [x] x87 FPU, MMX, SSE support
- [ ] AVX support
- [ ] `syscall/sysret` support (instead of the old `int 0x80` method for syscalling)

### Memory Management
- [ ] Copy-on-write support
- [ ] Support for swap/page backing storage.
- [ ] Rewrite `kheap`

### FAL and Filesystems
- [ ] Add a UNIX-like VFS.
- [ ] Clean up and improve the FAT driver:
  - [ ] Move the driver into a module.
  - [ ] Rewrite any bad or hard to understand code.
  - [ ] Add read support for FAT16 (`read`, `readdir`).
  - [ ] Add write support for FAT16/FAT32 (`write`, `mkdir`, `rmdir` etc).
- [ ] Add Ext2 support and use it for the system FS instead of FAT.
- [ ] Add `dev` filesystem. Should contain:
  - [ ] `tty*` device file(s) which user programs can use for terminal I/O.
  - [ ] `ttyS*` device fils(s) which user programs can use to access the serial port(s).
  - [ ] `fb` device file(s) that provide an interface between user programs and framebuffer(s).
  - [ ] `null` device
  - [ ] `zero` device
- [ ] Add `procfs` filesystem. Should contain:
  - [ ] Entries for all processes (including a `self` entry for the current process). These entries should contain info about the processes (idk what to put here. I'll try to do what Linux does).
  - [ ] `interrupts` file and maybe some other Linux procfs stuff too.
- [ ] Page cache
- [ ] Memory-mapped files

### DAL and Devices
- [ ] Char device abstraction
- [ ] Block device abstraction (maybe replace the drive thing with block devices?)

### Terminal Emulator
- [ ] Add support for (most) ANSI escape codes.
- [ ] Make the input system part of the terminal emulator.

### Scheduler and Userspace 
- [ ] Support for multiple threads per process (The mechanism for this should preferably be Linux-compatible)
- [ ] `fork()` Copy-on-write
- [ ] Process blocking/unblocking
- [ ] Multithreaded kernel
- [ ] Scheduler support for multiple processors (So actual SMP support)
- [ ] Pipes
- [ ] Signals
- [ ] As many POSIX and Linux syscalls as possible

### Userspace
- [ ]  Write a nicer shell (or port GNU Bash?)
- [ ] Implement some POSIX utilities (i.e. `cat`, `cp` etc)
- [ ] `reboot` and `shutdown` programs

### Security 
- [ ] Users and User Groups, UIDs and GIDs
- [ ] Login program

### Hardware
- [ ] IDE DMA Support
- [ ] PCIe Support

### Bug Fixes
- [ ] AHCI driver crashes the OS on real hardware.
- [ ] The OS breaks if the system partition (or any other partition for that matter?) is larger than ~512MB.
- [ ] scanf() seems to cause some wierd UB.
- [ ] mlibc sometimes panics when running user programs.

### Other
- [ ] Improve the build system (especially the way arch specific code is handled)
- [ ] Remove the kconfiglib python dependency by rewriting `genconfig.py` to use only standard functions.
- [ ] Fix that compiler warning hell.

This list is not complete, but it should give you an idea of what needs to be done
